[torchao][pt2e] Make prepare and convert faster by caching

navsud opened the pull request:
Summary: D79674759 tried to fix the expensive prepare and convert steps, as `assert_and_get_unique_device` was called multiple times. This change fixes that issue by using `functools.cache` decorator.

Test Plan:
Verified on llm export to QNN.
LLM Quantization prepare time of ~20min reduced to ~3min.

Rollback Plan:

Differential Revision: D82073679




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

andrewor14 reviewed (approved):
Looks great, thanks!

navsud committed:
[torchao][pt2e] Make prepare and convert faster by caching

Summary:
D79674759 tried to fix the expensive prepare and convert steps, as `assert_and_get_unique_device` was called multiple times. This change fixes that issue by using `functools.cache` decorator.

bypass-github-export-checks

Test Plan:
Verified on llm export to QNN.
Prepare time of ~20min reduced to ~3min.

Rollback Plan:

Reviewed By: andrewor14

Differential Revision: D82073679
